pkg/cmd/grafana/apiserver/auth/tokenAuthenticator.go

```diff
@@ -18,54 +25,169 @@ const (
 	verifier := authn.NewVerifier[CustomClaims](authn.IDVerifierConfig{
-		SigningKeysURL:   config.SigningKeysURL,
-		AllowedAudiences: config.AllowedAudiences,
+		SigningKeysURL:   options.IDVerifierConfig.SigningKeysURL,
+		AllowedAudiences: options.IDVerifierConfig.AllowedAudiences,
 	})
-	return getAccessTokenAuthenticatorFunc(&TokenValidator{verifier})
+
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	return &AccessTokenAuthenticator{
+		validator:      &TokenValidator{verifier},
+		serviceBaseURL: options.ServiceBaseURL,
+		systemCAPToken: options.SystemCAPToken,
+	}
 }
 
-func getAccessTokenAuthenticatorFunc(validator *TokenValidator) authenticator.RequestFunc {
-	return func(req *http.Request) (*authenticator.Response, bool, error) {
-		accessToken := req.Header.Get(headerKeyAccessToken)
+func newAuthAPIAuthorizedRequest(method, path string, token string, body []byte) (*http.Request, error) {
+	buffer := bytes.NewBuffer(nil)
+	if len(body) > 0 {
+		buffer.Write(body)
+	}
+
+	req, err := http.NewRequest(method, path, buffer)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+
+	return req, nil
+}
+
+func (auth *AccessTokenAuthenticator) signAccessToken(tokenValidationResult *authn.Claims[CustomClaims]) (string, error) {
+	body, err := json.Marshal(signAccessTokenRequest{})
+	if err != nil {
+		return "", err
+	}
+
+	path, err := url.JoinPath(auth.serviceBaseURL, "/v1/sign-access-token")
+	if err != nil {
+		return "", err
+	}
+	req, err := newAuthAPIAuthorizedRequest(http.MethodPost, path, auth.systemCAPToken, body)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add("X-Org-ID", tokenValidationResult.Rest.OrgId)
+
+	audience := strings.Split(tokenValidationResult.Claims.Audience[0], ":")
+	req.Header.Add("X-Realms", fmt.Sprintf(`[{"type":"%s","identifier":"%s"}]`, audience[0], audience[1]))
+
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+
+	var response signAccessTokenResponse
+	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
+		return "", err
+	}
+
+	if response.Status == "error" {
+		return "", fmt.Errorf(response.Error)
+	}
+
+	return response.AccessToken.Token, nil
+}
+
+func (auth *AccessTokenAuthenticator) AuthenticateRequest(req *http.Request) (*authenticator.Response, bool, error) {
+	var (
+		accessToken           string
+		tokenValidationResult *authn.Claims[CustomClaims]
+		err                   error
+	)
+
+	idToken := req.Header.Get(headerKeyGrafanaID)
+	if idToken == "" {
+		accessToken = req.Header.Get(headerKeyAccessToken)
 		if accessToken == "" {
 			return nil, false, nil
 		}
+		tokenValidationResult, err = auth.validator.Validate(req.Context(), accessToken)
+	} else {
+		tokenValidationResult, err = auth.validator.Validate(req.Context(), idToken)
+	}
```

Comment:
Access tokens and id tokens will have different claims attached to them. I have created a pr https://github.com/grafana/authlib/pull/25 to add wrapper for each type with their expected claims. Another change in that pr is that they will validate the `typ` header of a token so we cant reuse the same Validator for different kind of tokens.  I think its fine for now. When I get that pr merged I can update this part when we bump the dependency version

Reply:
I am moving this to enterprise, so as long as you/I can get that updated there.